### PR TITLE
Prelauncher

### PR DESCRIPTION
**Pull request by Laserlicht**
- Optional (enable setting in launcher options)
- Mainly to fix issue that android gamepad user needs to use mouse just for launching VCMI every time
  - now mouse is only needed for first launch and settings/mod-managment, but not for starting (if prelauncher is enabled)

Related: #4111 

grafik

### CODE REVIEW

File: launcher/settingsView/csettingsview_moc.ui [hunk 1]
```diff
@@ -1449,6 +1449,35 @@ Fullscreen Exclusive Mode - game will cover entirety of your screen and will use
          </attribute>
         </widget>
        </item>
+       <item row="61" column="0">
+        <widget class="QLabel" name="labelPreLauncher">
+         <property name="text">
+          <string>Prelauncher</string>
```
Comment (IvanSavenko) [inline launcher/settingsView/csettingsview_moc.ui:1455]: Prelauncher is very unclear IMO.
Not sure what to call it though. If somebody has better suggestions than this - go ahead.
```suggestion
          <string>Skip Launcher on start</string>
```

File: launcher/mainwindow_moc.cpp [hunk 2]
```diff
@@ -116,10 +116,43 @@ MainWindow::MainWindow(QWidget * parent)
...
 		UpdateDialog::showUpdateDialog(false);
 }
 
+void MainWindow::preLauncher()
+{
+	QMessageBox msgBox;
+	int secondsRemaining = 10;
+	QString text = tr("How to start?\n\nVCMI starts automatically in ");
+	QString textSeconds = tr(" Seconds");
+	msgBox.setText(text + QString::number(secondsRemaining) + textSeconds);
```
Comment (IvanSavenko) [inline launcher/mainwindow_moc.cpp:132]:
```suggestion
    QString textTemplate = tr("How to start?\n\nVCMI starts automatically in %1 seconds");
	msgBox.setText(textTemplate.arg(secondsRemaining));
```

Can be improved further - to allow plural translations (1 second, 2 seconds) but not sure about Qr syntax for that

**Review by IvanSavenko — COMMENTED** [review]
@kambala-decapitator any comments?

File: launcher/mainwindow_moc.cpp [hunk 3]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
 	if(settings["launcher"]["updateOnStartup"].Bool())
 		UpdateDialog::showUpdateDialog(false);
 }
 
+void MainWindow::preLauncher()
+{
+	QMessageBox msgBox;
+	int secondsRemaining = 10;
+	auto text = "How to start?\n\nVCMI starts automatically in %n second(s)";
+	msgBox.setText(tr(text, "", secondsRemaining));
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:131]: `tr` doesn't work with variables

```suggestion
	msgBox.setText(tr("How to start?\n\nVCMI starts automatically in %n second(s)", nullptr, secondsRemaining));
```

File: launcher/mainwindow_moc.cpp [hunk 4]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
+	msgBox.setWindowTitle(tr("VCMI"));
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:132]: why do you want to translate it?

File: launcher/mainwindow_moc.cpp [hunk 5]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
+	msgBox.setIcon(QMessageBox::Information);
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:133]:
```suggestion
	msgBox.setIcon(QMessageBox::Question);
```

File: launcher/mainwindow_moc.cpp [hunk 6]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
+	msgBox.addButton(tr("Start Launcher"), QMessageBox::YesRole);
+	QPushButton * startClient = msgBox.addButton(tr("Start VCMI"), QMessageBox::NoRole);
+	QTimer *timer = new QTimer(this);
+	connect(timer, &QTimer::timeout, this, [this, &timer, &text, &msgBox, &secondsRemaining](){
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:137]:
```suggestion
	connect(timer, &QTimer::timeout, this, [this, timer, &] {
```

File: launcher/mainwindow_moc.cpp [hunk 7]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
+		if(secondsRemaining == 0)
+		{
+			timer->stop();
+			msgBox.close();
+			ui->startGameButton->click();
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:142]: call the actual handler instead of simulating click

File: launcher/mainwindow_moc.cpp [hunk 8]
```diff
@@ -116,10 +116,42 @@ MainWindow::MainWindow(QWidget * parent)
...
+			return;
+		}
+		msgBox.setText(tr(text, "", secondsRemaining));
+		secondsRemaining--;
+	});
+	timer->start(1000);
+	msgBox.exec();
+	timer->stop();
+	if(msgBox.clickedButton() == startClient && secondsRemaining)
+		ui->startGameButton->click();
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:152]: same

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
- preLauncher is not a good name indeed
- this is needed only for Android, please enclose new functionality in ifdef

I don't really like the current implementation. Makes no sense to create MainWindow just to show the message box. As a quick solution it would do, but better make the message box a "starting" widget that would then create MainWindow if needed. For example, you could subclass QApplication and implement this logic there.

Ideally, the dialog should be implemented in Java and would never even start Qt activity if not needed. But that might not be easy to implement.

File: launcher/mainwindow_moc.cpp (hunk 3 above)
Comment (Laserlicht) [reply launcher/mainwindow_moc.cpp:131]: Then adding `tr("How to start?\n\nVCMI starts automatically in %n second(s)", nullptr, secondsRemaining)` twice in source?

Comment (kambala-decapitator) [reply launcher/mainwindow_moc.cpp:131]: better use lambda to create the string

File: launcher/mainwindow_moc.cpp [hunk 9]
```diff
@@ -116,10 +116,46 @@ MainWindow::MainWindow(QWidget * parent)
...
+
 	if(settings["launcher"]["updateOnStartup"].Bool())
 		UpdateDialog::showUpdateDialog(false);
 }
 
+void MainWindow::preLauncher()
+{
+	QMessageBox msgBox;
+	int secondsRemaining = 10;
+	auto setText = [&](){ msgBox.setText(tr("How to start?\n\nVCMI starts automatically in %n second(s)", "", secondsRemaining)); };
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:130]: it's better to just return a string from lambda and set it at the calling site:

```suggestion
	auto getText = [&] { return tr("How to start?\n\nVCMI starts automatically in %n second(s)", nullptr, secondsRemaining); };
```

File: launcher/mainwindow_moc.cpp [hunk 10]
```diff
@@ -116,10 +116,46 @@ MainWindow::MainWindow(QWidget * parent)
...
+	setText();
+	msgBox.setWindowTitle("VCMI");
+	msgBox.setIcon(QMessageBox::Question);
+	msgBox.addButton(tr("Start Launcher"), QMessageBox::YesRole);
+	QPushButton * startClient = msgBox.addButton(tr("Start VCMI"), QMessageBox::NoRole);
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:135]:
```suggestion
	auto startClientButton = msgBox.addButton(tr("Start VCMI"), QMessageBox::NoRole);
```

File: launcher/mainwindow_moc.cpp [hunk 11]
```diff
@@ -116,10 +116,46 @@ MainWindow::MainWindow(QWidget * parent)
...
+	QTimer *timer = new QTimer(this);
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:136]:
```suggestion
	auto timer = new QTimer(this);
```

File: launcher/mainwindow_moc.cpp [hunk 12]
```diff
@@ -116,10 +116,46 @@ MainWindow::MainWindow(QWidget * parent)
...
+	connect(timer, &QTimer::timeout, this, [&, this](){
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:137]:
```suggestion
	connect(timer, &QTimer::timeout, this, [this, timer, &] {
```

File: launcher/mainwindow_moc.cpp [hunk 13]
```diff
@@ -116,10 +116,46 @@ MainWindow::MainWindow(QWidget * parent)
...
+		if(secondsRemaining == 0)
+		{
+			timer->stop();
+			msgBox.close();
+			hide();
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:142]: does hiding improve something?

Comment (Laserlicht) [reply launcher/mainwindow_moc.cpp:142]: Hiding mouse for intro :)

File: launcher/mainwindow_moc.cpp [hunk 14]
```diff
@@ -127,14 +127,14 @@ void MainWindow::preLauncher()
...
-	setText();
+	auto getText = [&] { return tr("How to start?\n\nVCMI starts automatically in %n second(s)", nullptr, secondsRemaining); };
+	msgBox.setText(getText());
 	msgBox.setWindowTitle("VCMI");
 	msgBox.setIcon(QMessageBox::Question);
 	msgBox.addButton(tr("Start Launcher"), QMessageBox::YesRole);
-	QPushButton * startClient = msgBox.addButton(tr("Start VCMI"), QMessageBox::NoRole);
-	QTimer *timer = new QTimer(this);
-	connect(timer, &QTimer::timeout, this, [&, this](){
+	auto startClient = msgBox.addButton(tr("Start VCMI"), QMessageBox::NoRole);
```
Comment (kambala-decapitator) [inline launcher/mainwindow_moc.cpp:135]: please give proper names to variables: startClientButton

**Review by kambala-decapitator — COMMENTED** [review]
so, what about this?

> preLauncher is not a good name indeed
> this is needed only for Android, please enclose new functionality in ifdef

**Comment by IvanSavenko** [comment]
> portable gaming consoles (linux/windows -> there are some without touchscreen/touchpad out there)
pc (all OS) connected to TV

- Such systems most likely have mapping of controller input, e.g. sticks or D-pad to mouse input.
- You still need a way to launch VCMI itself, e.g. double-click on VCMI_Launcher.exe or any equivalent. So you should also be able to change whatever you need in Launcher and start client from it.
- On all PC OS's you can always start VCMI_Client directly for the same effect

What about hiding all this functionality behind a separate define, e.g. ENABLE_SKIP_LAUNCHER, and by default - enable it only on Android?

**Comment by Laserlicht** [comment]
>     * On all PC OS's you can always start VCMI_Client directly for the same effect

Yes you're right. Would be easier to add this one in player owned launcher.
Only our flatpak doesn't expose it seperatly atm (but should be addable over command line args).

> What about hiding all this functionality behind a separate define, e.g. ENABLE_SKIP_LAUNCHER, and by default - enable it only on Android?

What's the advantage over current solution? I think even most of android users are fine with starting in launcher.

**Comment by IvanSavenko** [comment]
> What's the advantage over current solution? I think even most of android users are fine with starting in launcher.

I meant in addition to current solution.

So this option will be available only on Android (where there is known use-case for it), and toggled to false by default. While for all other platforms it will be disabled / not compiled in. 

And if we decide that another platform needs it or we want to enable it everywhere we can adjust that single define instead of searching for all `#ifdef VCMI_ANDROID` related to this feature.

**Comment by Laserlicht** [comment]
> I meant in addition to current solution.

Yeah, but still don't understand the point. :D

Adding code to save maybe 100 Bytes in binary and to disallow usage of an already optional feature for non android users?

I mean. This feature is also working on PC. Maybe also someone prefer it like this. And if not - there is also no problem.

**Comment by kambala-decapitator** [comment]
> Yes we could do a separate window/view. But I think it would be overengineered for this little functionality.

writing clean code may often feel like overengineering, but in the end you get benefits :)

> What about hiding all this functionality behind a separate define, e.g. ENABLE_SKIP_LAUNCHER, and by default - enable it only on Android?

sounds good. IMO this feature is useless everywhere (for reasons already given) except Android.

**Comment by IvanSavenko** [comment]
> Yeah, but still don't understand the point. :D

The way I see it, this is a workaround for an Android-specific issue that (hopefully) will be unnecessary and safe to remove once somebody adds a way to use controller in Launcher on Android. And not an independent feature.

So from this point of view, having it Android-specific makes sense.

**Pull request closed without merging**